### PR TITLE
python3Packages.ruff-lsp: init at 0.0.24

### PR DESCRIPTION
**Pull request by kalekseev**
###### Description of changes

Init https://github.com/charliermarsh/ruff-lsp package, there's another pr but it doesn't work for me and no activity for two months https://github.com/NixOS/nixpkgs/pull/207250
I'll be happy with any of the PRs merged.
cc @cpcloud @fabaff 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/ruff-lsp/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    hatchling
+  ];
+
+  propagatedBuildInputs = [
+    pygls
+    typing-extensions
+  ];
+
+  postPatch = ''
+    sed -i '/"ruff>=/d' pyproject.toml
```
Comment (fabaff) [inline pkgs/development/python-modules/ruff-lsp/default.nix:33]: Upstream mention `ruff` as [run-time](https://github.com/charliermarsh/ruff-lsp/blob/main/pyproject.toml#L38) dependency.

File: pkgs/development/python-modules/ruff-lsp/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    sed -i 's|USER_DEFAULTS: dict\[str, str\] = {}|USER_DEFAULTS: dict[str, str] = {"path": ["${ruff}/bin/ruff"]}|' ruff_lsp/server.py
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/charliermarsh/ruff-lsp";
+    description = "A Language Server Protocol implementation for Ruff";
```
Comment (fabaff) [inline pkgs/development/python-modules/ruff-lsp/default.nix:39]: The changelog will be automatically attached to the commit messages by tool (e.g., `nix-update`).

```suggestion
    description = "A Language Server Protocol implementation for Ruff";
    changelog = "https://github.com/charliermarsh/ruff-lsp/releases/tag/v${version}";
```

File: pkgs/development/python-modules/ruff-lsp/default.nix (hunk 1 above)
Comment (kalekseev) [reply pkgs/development/python-modules/ruff-lsp/default.nix:33]: ruff on pypi just installs ruff binary (written in rust) and then this library search for it on filesystem, in next sed I pass that binary as default

Comment (kalekseev) [reply pkgs/development/python-modules/ruff-lsp/default.nix:43]: ruff-lsp now searches for ruff binary in PATH, added ruff from nixpkgs to wrapper

**Comment by mweinelt** [comment]
There is also #216841.

**Comment by loicreynier** [comment]
`ruff-lsp` can be updated to 0.0.19 before merging.

**Comment by kalekseev** [comment]
@fabaff could you review this please?

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 214720` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.ruff-lsp</li>
    <li>python310Packages.ruff-lsp.dist</li>
    <li>python311Packages.ruff-lsp</li>
    <li>python311Packages.ruff-lsp.dist</li>
  </ul>
</details>

Tests are failing:
```
       > +                  {'code': 'F821',
       > +                   'data': {'fix': None, 'noqa_row': 3},
       > +                   'message': 'Undefined name `x`',
       > +                   'range': {'end': {'character': 7, 'line': 2},
       > +                             'start': {'character': 6, 'line': 2}},
       > +                   'severity': 1,
       > +                   'source': 'Ruff'}],
       > - {'diagnostics': [], 'uri': 'file:///build/tmp7_ngkmwz.py'}
       > ? -------------------
       >
       > +  'uri': 'file:///build/tmp7_ngkmwz.py'}
       >
       > ----------------------------------------------------------------------
       > Ran 2 tests in 0.828s
       >
       > FAILED (failures=2)
       > /nix/store/5s1yg5l36wzgy1dj0vv1ibarc4g7vrdr-stdenv-linux/setup: line 1594: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/sf96dvqm6k82hz8qvfpq8cdhvamsr76s-python3.10-ruff-lsp-0.0.24.drv'.
error: builder for '/nix/store/2y40r03xspznxfqr8pd968i6i3mvn1zf-python3.11-ruff-lsp-0.0.24.drv' failed with exit code 1;
+                   'range': {'end': {'character': 10, 'line': 0},
+                             'start': {'character': 7, 'line': 0}},
+                   'severity': 2,
+                   'source': 'Ruff',
+                   'tags': [1]},
+                  {'code': 'F821',
+                   'data': {'fix': None, 'noqa_row': 3},
+                   'message': 'Undefined name `x`',
+                   'range': {'end': {'character': 7, 'line': 2},
+                             'start': {'character': 6, 'line': 2}},
+                   'severity': 1,
+                   'source': 'Ruff'}],
- {'diagnostics': [], 'uri': 'file:///build/tmp7mmxyo8r.py'}
? -------------------

+  'uri': 'file:///build/tmp7mmxyo8r.py'}

======================================================================
FAIL: test_no_initialization_options (tests.test_server.TestServer.test_no_initialization_options)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/ruff_lsp-0.0.24/tests/test_server.py", line 181, in test_no_initialization_options
    self.assertEqual(actual, expected)
AssertionError: {'uri[46 chars]s': []} != {'uri[46 chars]s': [{'range': {'start': {'line': 0, 'characte[533 chars]f'}]}
+ {'diagnostics': [{'code': 'F401',
+                   'data': {'fix': {'edits': [{'content': '',
+                                               'end_location': {'column': 0,
+                                                                'row': 2},
+                                               'location': {'column': 0,
+                                                            'row': 1}}],
+                                    'message': 'Remove unused import: `sys`'},
+                            'noqa_row': 1},
+                   'message': '`sys` imported but unused',
+                   'range': {'end': {'character': 10, 'line': 0},
+                             'start': {'character': 7, 'line': 0}},
+                   'severity': 2,
+                   'source': 'Ruff',
+                   'tags': [1]},
+                  {'code': 'F821',
+                   'data': {'fix': None, 'noqa_row': 3},
+                   'message': 'Undefined name `x`',
+                   'range': {'end': {'character': 7, 'line': 2},
+                             'start': {'character': 6, 'line': 2}},
+                   'severity': 1,
+                   'source': 'Ruff'}],
- {'diagnostics': [], 'uri': 'file:///build/tmppxpuymvg.py'}
? -------------------

+  'uri': 'file:///build/tmppxpuymvg.py'}

----------------------------------------------------------------------
Ran 2 tests in 0.722s

FAILED (failures=2)
```

**Comment by wegank** [comment]
@ofborg build python3Packages.ruff-lsp python3Packages.ruff-lsp.passthru.tests

**Comment by kalekseev** [comment]
Result of `nixpkgs-review pr 214720` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.ruff-lsp</li>
    <li>python310Packages.ruff-lsp.dist</li>
    <li>python311Packages.ruff-lsp</li>
    <li>python311Packages.ruff-lsp.dist</li>
  </ul>
</details>

**Comment by kalekseev** [comment]
Will investigate linux build when I have time

**Comment by GaetanLepage** [comment]
I tested locally and I don't understand why the test is failing.
The ruff binary is clearly present.
Running
```
ruff --version
python -m ruff_lsp --version
```
in the `preCheck` phase works fine.

However, when running the LSP from the `test_server.py` file to analyze
```python
import sys

print(x)
```
no diagnostic is returned, so the test fails.

It is even weirder if the same test succeeds on your mac.

Note: Running the test manually from the repo works fine.
```
git clone https://github.com/charliermarsh/ruff-lsp.git
cd ruff-lsp
python -m venv venv
source venv/bin/activate
just install
just test
```

**Comment by wegank** [comment]
`nix-build -A python3Packages.ruff-lsp --option sandbox false` does succeed on my aarch64-linux machine.

**Comment by GaetanLepage** [comment]
> `nix-build -A python3Packages.ruff-lsp --option sandbox false` does succeed on my aarch64-linux machine. So, disable the tests for now?

The same command is failing on my `x86_64-linux` machine.
However, the ofborg build seems to fail for both `x86_64-linux` and `aarch64-linux`.
Do you know why @wegank ?

I don't mind disabling the tests, but it might be worth it to test it in real life condition then.

**Comment by wegank** [comment]
I thought it was a sandbox issue, since sandbox is usually enabled on Linux and disabled on macOS. If this is the case, I guess there's nothing we can do in the Nix expressions to work around the failure.

**Comment by GaetanLepage** [comment]
> I thought it was due to sandbox, since sandbox is usually enabled on Linux and disabled on macOS. Which implies, for example, that network access is only disabled under Linux.
> 
> If this is the case, I guess there's nothing we can do in the Nix expressions to work around the failure.

Does it work for you without disabling sandbox ?

**Comment by wegank** [comment]
> Does it work for you without disabling sandbox ?

Nope.

**Comment by GaetanLepage** [comment]
> Nope.

I tested on two `x86_64-linux` systems (one is NixOS, the other one is Ubuntu 22.04) and the tests fail on both systems with and without sandbox...

**Comment by GaetanLepage** [comment]
Well @kalekseev, we might then want to add
`doCheck = stdenv.isDarwin;` then. Or maybe disable them completely.

**Comment by kalekseev** [comment]
> Well @kalekseev, we might then want to add
> `doCheck = stdenv.isDarwin;` then. Or maybe disable them completely.

I'll review and update today or tomorrow, thanks for your help!

**Comment by kalekseev** [comment]
Pushed `doCheck=isDarwin`.
On my aarch64-linux tests fails in build but passes in nix-shell so I guess it some env problem, not a broken build.

```
[nix-shell:~/nixpkgs/ruff_lsp-0.0.24]$ unittestCheckPhase 
Executing unittestCheckPhase
..
----------------------------------------------------------------------
Ran 2 tests in 0.516s

OK
Finished executing unittestCheckPhase
```

File: pkgs/development/python-modules/ruff-lsp/default.nix [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  propagatedBuildInputs = [
+    pygls
+    lsprotocol
+    typing-extensions
+  ];
+
+  postPatch = ''
+    # ruff binary added to PATH in wrapper so it's not needed
+    sed -i '/"ruff>=/d' pyproject.toml
+  '';
```
Comment (wegank) [inline pkgs/development/python-modules/ruff-lsp/default.nix:47]: This should be right under `src`.

File: pkgs/development/python-modules/ruff-lsp/default.nix [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  nativeBuildInputs = [
+    hatchling
+  ];
+
+
+  nativeCheckInputs = [
+    unittestCheckHook
+    python-lsp-jsonrpc
+    ruff
+  ];
```
Comment (wegank) [inline pkgs/development/python-modules/ruff-lsp/default.nix:36]: And this under `doCheck`.

Comment (kalekseev) [reply pkgs/development/python-modules/ruff-lsp/default.nix:36]: Are these requests personal preferences or documented somewhere as best practices?

**Pull request merged**